In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

df_police = pd.read_csv('Police_Data.csv')
df_police.head()

ID Reference Number           Subject's race Subject's gender  \
0                     1                   Asian             Male   
1                     2  Black/African American             Male   
2                     3  Black/African American             Male   
3                     4  Black/African American             Male   
4                     5                   White           Female   

   Subject's age     Ethnicity  District of occurrence  Adjacent to School  \
0             43  NON-HISPANIC                     6.0                   0   
1             23  NON-HISPANIC                     1.0                   0   
2             18  NON-HISPANIC                     1.0                   0   
3             21  NON-HISPANIC                     1.0                   0   
4             38      HISPANIC                     4.0                   0   

  Assigned Division Assigned Bureau Event Date/Time  
0               TOD             FSB   1/1/2021 1:07  
1              RCPD             NaN   1/1/2021 0:52  
2              RCPD             NaN   1/1/2021 0:52  
3              RCPD             NaN   1/1/2021 0:52  
4               TOD             FSB   1/1/2021 3:00

In [3]:
df_police.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23510 entries, 0 to 23509
Data columns (total 10 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   ID Reference Number     23510 non-null  int64  
 1   Subject's race          23510 non-null  object 
 2   Subject's gender        23510 non-null  object 
 3   Subject's age           23510 non-null  int64  
 4   Ethnicity               23510 non-null  object 
 5   District of occurrence  23408 non-null  float64
 6   Adjacent to School      23510 non-null  int64  
 7   Assigned Division       23381 non-null  object 
 8   Assigned Bureau         18851 non-null  object 
 9   Event Date/Time         23510 non-null  object 
dtypes: float64(1), int64(3), object(6)
memory usage: 1.8+ MB


In [4]:
df_police.shape

(23510, 10)

In [5]:
df_police.describe()

ID Reference Number   Subject's age  District of occurrence  \
count           23510.00000   23510.000000            23408.000000   
mean            11755.50000      31.813143                3.989192   
std              6786.89675      12.946909                1.748587   
min                 1.00000       1.000000                1.000000   
25%              5878.25000      22.000000                3.000000   
50%             11755.50000      30.000000                4.000000   
75%             17632.75000      39.000000                5.000000   
max             23510.00000      99.000000                8.000000   

       Adjacent to School  
count        23510.000000  
mean             0.085155  
std              0.279119  
min              0.000000  
25%              0.000000  
50%              0.000000  
75%              0.000000  
max              1.000000

In [6]:
df_police[df_police['District of occurrence'].isnull()]

ID Reference Number           Subject's race Subject's gender  \
145                     146                   White             Male   
678                     679  Black/African American             Male   
1005                   1006  Black/African American             Male   
2991                   2992  Black/African American             Male   
2992                   2993                   White             Male   
...                     ...                     ...              ...   
22101                 22102                   Asian             Male   
22216                 22217  Black/African American             Male   
22217                 22218  Black/African American             Male   
22218                 22219                   White             Male   
23231                 23232  Black/African American           Female   

       Subject's age     Ethnicity  District of occurrence  \
145               24      HISPANIC                     NaN   
678               37  NON-HISPANIC                     NaN   
1005              56  NON-HISPANIC                     NaN   
2991              26  NON-HISPANIC                     NaN   
2992              50  NON-HISPANIC                     NaN   
...              ...           ...                     ...   
22101             24  NON-HISPANIC                     NaN   
22216             14  NON-HISPANIC                     NaN   
22217             14  NON-HISPANIC                     NaN   
22218             14  NON-HISPANIC                     NaN   
23231             55  NON-HISPANIC                     NaN   

       Adjacent to School Assigned Division Assigned Bureau  Event Date/Time  
145                     0              TPPD             NaN   1/9/2021 23:16  
678                     0                3D             PSB   2/16/2021 0:00  
1005                    0                2D             PSB  3/11/2021 13:42  
2991                    0               GPD             NaN  7/26/2021 23:28  
2992                    0                3D             PSB   6/3/2021 11:48  
...                   ...               ...             ...              ...  
22101                   0                2D             PSB  4/17/2024 10:57  
22216                   0                4D             PSB  4/22/2024 15:00  
22217                   0                4D             PSB  4/22/2024 15:00  
22218                   0                4D             PSB  4/22/2024 15:00  
23231                   0              TPPD             NaN  6/12/2024 16:38  

[102 rows x 10 columns]

In [14]:
df_police['District of occurrence'].value_counts(dropna= False, normalize= True)

District of occurrence
3.0    0.203403
6.0    0.201957
4.0    0.194088
5.0    0.162484
1.0    0.106891
2.0    0.093407
8.0    0.033433
NaN    0.004339
Name: proportion, dtype: float64

In [8]:
df_police[df_police['Assigned Division'].isnull()]

ID Reference Number           Subject's race Subject's gender  \
542                     543                   White             Male   
688                     689  Black/African American           Female   
689                     690  Black/African American             Male   
831                     832  Black/African American             Male   
930                     931  Black/African American             Male   
...                     ...                     ...              ...   
21236                 21237                   White           Female   
21237                 21238                   White           Female   
21238                 21239                   White           Female   
21326                 21327                   White           Female   
23202                 23203                   White           Female   

       Subject's age     Ethnicity  District of occurrence  \
542               43      HISPANIC                     3.0   
688               21  NON-HISPANIC                     1.0   
689               27  NON-HISPANIC                     1.0   
831               49  NON-HISPANIC                     5.0   
930               39  NON-HISPANIC                     1.0   
...              ...           ...                     ...   
21236             17      HISPANIC                     4.0   
21237             16      HISPANIC                     4.0   
21238             16      HISPANIC                     4.0   
21326             13      HISPANIC                     4.0   
23202             16      HISPANIC                     4.0   

       Adjacent to School Assigned Division Assigned Bureau  Event Date/Time  
542                     0               NaN             NaN   2/8/2021 10:44  
688                     0               NaN             NaN  2/18/2021 18:20  
689                     0               NaN             NaN  2/18/2021 18:20  
831                     0               NaN             NaN   3/3/2021 10:50  
930                     0               NaN             NaN   3/8/2021 12:55  
...                   ...               ...             ...              ...  
21236                   0               NaN             NaN   3/6/2024 13:20  
21237                   0               NaN             NaN   3/6/2024 13:20  
21238                   0               NaN             NaN   3/6/2024 13:20  
21326                   1               NaN             NaN   3/7/2024 15:00  
23202                   0               NaN             NaN   6/7/2024 12:41  

[129 rows x 10 columns]

In [15]:
df_police['Assigned Division'].value_counts(dropna= False, normalize= True)

Assigned Division
3D      0.184219
4D      0.166227
5D      0.148958
GPD     0.102254
2D      0.079541
6D      0.075117
RCPD    0.054530
TOD     0.043939
1D      0.039387
TPPD    0.033262
IMTD    0.019694
CED     0.015355
SID     0.010804
SVID    0.010591
NaN     0.005487
MCD     0.003190
CID     0.003148
MCFM    0.002084
MCSO    0.000808
PSTA    0.000723
SOD     0.000298
FSB     0.000170
ISB     0.000128
PSCC    0.000043
PERS    0.000043
Name: proportion, dtype: float64

In [18]:
df_police['Assigned Bureau'].value_counts(dropna= False)

Assigned Bureau
PSB    16309
NaN     4659
FSB     1405
ISB      655
MSB      482
Name: count, dtype: int64

District of Occurrence has lots of missing values, 102 to be exact. However, this is less than a 10th of a percent. Dropping these rows might be best. Similarly, Assigned Division also has NaNs that are fairly low in comparison. Assigned Bureau however, has 4659 missing values. This is important to bring up, as well as might require actual imputing instead of just dropping.

For now, let's go ahead and convert the datetime column to an actual datetime object. We should also set some columns to the categorical type, and clean the column names just in case.

In [23]:
df_police.columns = df_police.columns.str.strip()
df_police['Event Date/Time'] = pd.to_datetime(df_police['Event Date/Time'])
categoricals = ['ID Reference Number', 'Subject\'s race', 'Subject\'s gender', 'Ethnicity', 'District of occurrence']
df_police[categoricals] = df_police[categoricals].astype('category')

In [25]:
df_police['District of occurrence'].dtype


CategoricalDtype(categories=[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 8.0], ordered=False, categories_dtype=float64)